### DISCUSSION TITLE
Does GetStringMap work with ENV variables?

### DISCUSSION THREAD

**Discussion by treddy73** [spf13/viper#1123] [Q&A]
I'm currently using a YAML file for my configuration and it has a structure like:
```
db:
  debug: true
  driverName: postgres
  dataSourceName:
    dbname: thedb
    user: theuser
    password: thepassword
    host: localhost
    port: 5432
    sslmode: disable
```
`v.GetStringMap("db.dataSourceName")` returns a map that contains all of the child keys and it seems to work the way I want it to.  I do this to help programmatically build the data source name.

However, if I use environment variables like:
```
DB_DATASOURCENAME_DBNAME=thedb
DB_DATASOURCENANE_USER=theuser
etc ...
```
and I additionally configure Viper:
```
v.AutomaticEnv()
v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
```
then `v.GetStringMap("db.dataSourceName")` returns `nil`.

I am able to get the individual keys directly `v.GetString("db.dataSourceName.dbname")` for instance.

Am I unable to grab a hierarchy of keys from an ENV configuration?

**Comment by mmsqe** [comment]
@treddy73 seems it's supported like 
```
type Config struct {
	DB struct {
		DataSourceName map[string]interface{}
	}
}

func main() {
	var cfg = []byte(`
		{
			"db": {
				"data-source-name": {
					"user": "theuser",
					"dbname": "thedb"
				}
			}
		}
	`)
	viper.SetConfigType("json")
	viper.ReadConfig(bytes.NewBuffer(cfg))
	fmt.Println(viper.GetStringMap("db.data-source-name"))
	fmt.Println(viper.GetString("db.data-source-name.user"))
}
```

**Reply by sflanker** [reply]
The OP is asking about environment variables specifically. Your example is purely loading settings from configuration.

**Comment by sflanker** [comment]
I don't believe Viper has any support for this scenario. If you look at GetStringMap it is just a wrapper for the `find` function which will only return a `map` if the value is from a config file or K/V store. The environment variable paths can only return a string. So the only way to use environment variables with GetStringMap would be to pack the keys into a json object:

```
DB_DATASOURCENAME='{ "dbname": "thedb", "dbuser": "theuser" }' ...
```

If you ask me this is a serious limitation in Viper and makes it a bit lacking compared to offerings for other programming languages such as .Net or Node.js.
